src/battle-log.ts

```diff
@@ -93,16 +99,37 @@ class BattleLog {
 			}
 			break;
 
-		case 'join': case 'j': {
+		case 'join': case 'j': case 'leave': case 'l': {
 			const user = BattleTextParser.parseNameParts(args[1]);
-			divHTML = '<small>' + BattleLog.escapeHTML(user.group + user.name) + ' joined.</small>';
-			break;
-		}
-		case 'leave': case 'l': {
-			const user = BattleTextParser.parseNameParts(args[1]);
-			divHTML = '<small>' + BattleLog.escapeHTML(user.group + user.name) + ' left.</small>';
+			const formattedUser = BattleLog.escapeHTML(user.group + user.name);
+			const isJoin = args[0].includes('j');
+			if (!this.joinLeave) {
+				this.joinLeave = {
+					joins: [],
+					leaves: [],
+					element: document.createElement('div') as HTMLDivElement,
+				};
+			}
+
+			if (isJoin && this.joinLeave.leaves.includes(formattedUser)) {
+				this.joinLeave.leaves.splice(this.joinLeave.leaves.indexOf(formattedUser), 1);
+			} else {
+				this.joinLeave[isJoin ? "joins" : "leaves"].push(formattedUser);
+			}
+
+			let buf = '';
+			if (this.joinLeave.joins.length) {
+				buf += this.displayJoinLeaves(this.joinLeave.joins, 'joined');
+			}
+			if (this.joinLeave.leaves.length) {
+				if (this.joinLeave.joins.length) buf += '; ';
+				buf += this.displayJoinLeaves(this.joinLeave.leaves, 'left') + '<br />';
+			}
+			this.joinLeave.element.innerHTML = `<small style="color: #555555">${buf}</small>`;
```

Comment:
```suggestion
			this.joinLeave.element.innerHTML = `<small>${buf}</small>`;
```

(The style isn't necessary if you're setting the class to `chat`.)